Extend ST_Points to support major Well-Known spatial objects

uprush opened the pull request:
Extend ST_Points to support POINT, LINESTRING, POLYGON, MULTIPOINT,
MULTILINESTRING, MULTIPOLYGON and GEOMETRYCOLLECTION.

```
== RELEASE NOTES ==

Geospatial Changes
* Improve ST_Points to add support for major Well-Known spatial objects.
  ST_Points now supports POINT, LINESTRING, POLYGON, MULTIPOINT, MULTILINESTRING, MULTIPOLYGON and GEOMETRYCOLLECTION.
```

rschlussel commented:
@jagill

jagill commented:
Thanks @uprush !  First, I really appreciate the good tests, and I think the direction is right. However, I think we should no de-duplicate points, like this code does for Polygons.  My thoughts are below.

1. ST_Points is only defined in the ISO standard for LineStrings.  In this case, it's explicit that ST_Points starts and ends with the first and last point, even if they are the same -- ie, a closed loop.  This is what's currently implemented.  This suggests that duplicate points are ok.
2. Extending ST_Points to Point and MultiPoint seems non-ambiguous.
3. For MultiLineString, it seems right to concatenate the results of ST_Point from the constituent LineStrings, in order.  But what if some of the boundary points are the same?  This is a valid MLS; I think in this case we should not de-duplicate them.  Could we put a test for this?
4. For Polygon, I think the right way to view this is as a MultiLineString containing the exterior loop and interior loops (in the natural order).  Then we can use the rule 3 above.  Current the resultant points don't have the end point of the loops.  Also, the order seems strange, but this could happen naturally due to the geometries re-orienting polygon loops after they are parsed from WKT.
5. For MultiPolygons and GeometryCollections, we should concatenate the results of the constituents, ideally in order.

uprush commented:
@jagill, thanks for your review. Following your comments, I have changed the code to not de-duplicate the points. This also makes the Polygon a closed-loop. I also refactored my code to use the JST Geometry instead of OSGGeometry, which should preserve the order of the constituents of the results in MultiPolygons and GeometryCollections.

uprush commented:
Hi @jagill, any update on the review?

jagill commented on presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java:
```diff
@@ -868,57 +870,38 @@ public static Slice stEndPoint(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
+    private static void buildPointsBlock(Geometry geometry, BlockBuilder blockBuilder)
     {
-        GeometryType type = GeometryType.getForEsriGeometryType(geometry.geometryType());
+        GeometryType type = GeometryType.getForEsriGeometryType(geometry.getGeometryType());
```
If this is a JTS geometry now, this should use `getForJtsGeometryType`

uprush replied:
Got it.

jagill commented on presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java:
```diff
@@ -868,57 +870,38 @@ public static Slice stEndPoint(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
-            MultiVertexGeometry vertices = (MultiVertexGeometry) geometry.getEsriGeometry();
-            for (int i = 0; i < vertices.getPointCount(); i++) {
-                GEOMETRY.writeSlice(blockBuilder, serialize(createFromEsriGeometry(vertices.getPoint(i), null)));
+            GeometryFactory gFactory = geometry.getFactory();
```
Instead of abbreviating, use `geometryFactory`.

uprush replied:
Done.

jagill commented:
Great!  Could you:
1. Squash the two commits into one, and
2. Rebase to master?

It'll be easier to understand the changes when they aren't split into two intersecting commits with the same commit message.

uprush commented:
Thanks. The commits have been squashed and rebased.

jagill commented on presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java:
```diff
@@ -862,24 +868,46 @@ public static Slice stEndPoint(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
     {
         Geometry geometry = deserialize(input);
-        validateType("ST_Points", geometry, EnumSet.of(LINE_STRING));
+        validateType("ST_Points", geometry, VALID_TYPES_FOR_ST_POINTS);
```
Since this is all current geometries, and the method would work for any additional geometry that JTS supports, we can remove this (and the associated bit in lines 150-152.

uprush replied:
Sounds great. Removing this check.

jagill commented on presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java:
```diff
@@ -862,24 +868,46 @@ public static Slice stEndPoint(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
-            GEOMETRY.writeSlice(blockBuilder, serialize(lineString.getPointN(i)));
-        }
+
+        int pointCount = geometry.getNumPoints();
```
You can inline the pointCount variable, since it's only used in one place:
```java
BlockBuilder blockBuilder = GEOMETRY.createBlockBuilder(null, geometry.getNumPoints());
```

uprush replied:
Thanks @jagill. I removed pointCount.

jagill commented:
@rschlussel Would you want to take a look at this?

rschlussel commented on presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java:
```diff
@@ -601,7 +601,47 @@ public void testSTPoints()
+
+        assertFunction("ST_Points(ST_GeometryFromText('POLYGON EMPTY'))", new ArrayType(GEOMETRY), null);
+        assertSTPoints("POLYGON ((8 4, 3 9, 5 6, 8 4))", "8 4", "5 6", "3 9", "8 4");
+        assertSTPoints("POLYGON ((8 4, 3 9, 5 6, 7 2, 8 4))", "8 4", "7 2", "5 6", "3 9", "8 4");
```
i'm curious why the order changes and if the order matters

jagill replied:
A polygon's shell has an ISO specified order (counter-clockwise), but it can accept vertices in either order (and will convert them upon ingestion).  To complicate matters further, one of the main geometry libraries (ESRI) converts polygon shells to the opposite order (clockwise).

rschlussel commented on presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java:
```diff
@@ -601,7 +601,47 @@ public void testSTPoints()
+
+        assertFunction("ST_Points(ST_GeometryFromText('GEOMETRYCOLLECTION EMPTY'))", new ArrayType(GEOMETRY), null);
+        String newLine = System.getProperty("line.separator");
+        String geometryCollection = String.join(newLine,
```
why do the newlines matter?

jagill replied:
Newlines don't matter for Java parsing (although some strict parsers don't allow newlines).

uprush commented:
I accidentally resolved the conversation about points order. If no matter what order in the WKT, the ESRI library will always convert it to clockwise. How about we changed the order in the test WKTs to match the converted order like the below to avoid confusion. I made the change locally, test passed.

```
         assertFunction("ST_Points(ST_GeometryFromText('MULTIPOLYGON EMPTY'))", new ArrayType(GEOMETRY), null);
-        assertSTPoints("MULTIPOLYGON (((0 0, 4 0, 4 4, 0 4, 0 0), (1 1, 2 1, 2 2, 1 2, 1 1)), ((-1 -1, -1 -2, -2 -2, -2 -1, -1 -1)))",
+        assertSTPoints("MULTIPOLYGON (((0 0, 0 4, 4 4, 4 0, 0 0), (1 1, 2 1, 2 2, 1 2, 1 1)), ((-1 -1, -1 -2, -2 -2, -2 -1, -1 -1)))",
                 "0 0", "0 4", "4 4", "4 0", "0 0",
                 "1 1", "2 1", "2 2", "1 2", "1 1",
                 "-1 -1", "-1 -2", "-2 -2", "-2 -1", "-1 -1");
```

rschlussel commented:
Thanks for the explanation about the clockwise order.  No need to change the order in the test.  Just update to remove the newline and we should be good to go.

uprush commented:
@rschlussel @jagill are we good to merge? There was a build failure, but shouldn't be related to my last commit since I only removed the newline.

jagill commented:
@uprush I agree the test failures look spurious.  Can you rebase onto the
latest master and push, which will trigger another job and hopefully avoid
the failures?

On Wed, Feb 12, 2020 at 4:30 AM Yifeng Jiang <notifications@github.com>
wrote:

> @rschlussel <https://github.com/rschlussel> @jagill
> <https://github.com/jagill> are we good to merge? There was a build
> failure, but shouldn't be related to my last commit since I only removed
> the newline.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/prestodb/presto/pull/13823?email_source=notifications&email_token=AADZAMDXR2RJ5AWZLMOLQ2LRCO6VBA5CNFSM4JYF5SFKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOELQCDWQ#issuecomment-585114074>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADZAMG6D2353DK4X6LVD4LRCO6VBANCNFSM4JYF5SFA>
> .
>

uprush commented:
@jagill Okay. Just rebased on latest master and pushed.
